File: src/Billing/Api/Internal/ProcessCdrBatchHandler.cs
using System.Text.Json;
using Billing.Service.Beans;
using Grpc.Core;
using Telcobright.Billing.V1;

namespace Billing.Service.Api.Internal;

/// <summary>
/// Adapts the gRPC <c>ProcessCdrBatch</c> RPC to the <see cref="CdrProcessor"/> bean: deserialize each cdr
/// from its JSON (the Kafka payload shape) into the full <c>cdr</c> POCO, hand the batch to the processor
/// (which resolves the tenant's config from config-manager and mediates + writes it), then map the outcome
/// back to the proto reply. All the processing lives in the bean; this is just the proto boundary.
/// </summary>
public sealed class ProcessCdrBatchHandler
{
    private static readonly JsonSerializerOptions CdrJson = new() { PropertyNameCaseInsensitive = true };

    private readonly CdrProcessor _cdrProcessor;

    public ProcessCdrBatchHandler(CdrProcessor cdrProcessor) => _cdrProcessor = cdrProcessor;

    public Task<CdrBatchResult> Handle(CdrBatchRequest request, ServerCallContext context)
    {
        List<MediationModel.cdr> cdrs;
        try
        {
            cdrs = new List<MediationModel.cdr>(request.CdrsJson.Count);
            foreach (var json in request.CdrsJson)
                cdrs.Add(JsonSerializer.Deserialize<MediationModel.cdr>(json, CdrJson)!);
        }
        catch (Exception ex)
        {
            return Task.FromResult(new CdrBatchResult { Error = "cdr json parse error: " + ex.Message });
        }

        var result = _cdrProcessor.ProcessBatch(request.Tenant, cdrs);
        if (result.Batch is null)
            return Task.FromResult(new CdrBatchResult { Committed = result.Committed, Error = result.Error ?? "" });

        var b = result.Batch;
        return Task.FromResult(new CdrBatchResult
        {
            Committed = true,
            Rated = b.Rated.Count,
            Errored = b.Errored.Count,
            CdrsWritten = b.CdrsWritten,
            ChargeablesWritten = b.ChargeablesWritten,
            CdrErrorsWritten = b.CdrErrorsWritten,
            TotalCharged = (double)b.TotalCharged,
        });
    }
}

File: java/src/main/java/com/telcobright/billing/api/internal/ProcessCdrBatchHandler.java
package com.telcobright.billing.api.internal;

import com.fasterxml.jackson.databind.DeserializationFeature;
import com.fasterxml.jackson.databind.MapperFeature;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.telcobright.billing.beans.CdrProcessingResult;
import com.telcobright.billing.beans.CdrProcessor;
import com.telcobright.billing.grpc.CdrBatchRequest;
import com.telcobright.billing.grpc.CdrBatchResult;
import com.telcobright.billing.mediation.engine.models.cdr;
import jakarta.inject.Inject;
import jakarta.inject.Singleton;

import java.util.ArrayList;
import java.util.List;

/**
 * Adapts the gRPC {@code ProcessCdrBatch} RPC to the {@link CdrProcessor} bean: deserialize each cdr from its
 * JSON (the Kafka payload shape) into the full {@code cdr} POCO, hand the batch to the processor (which
 * resolves the tenant's config from config-manager and mediates + writes it), then map the outcome back to
 * the proto reply. All the processing lives in the bean; this is just the proto boundary.
 */
@Singleton
public class ProcessCdrBatchHandler {
    // case-insensitive, ignore-unknown, JSR-310 dates — mirrors the .NET JsonSerializerOptions{PropertyNameCaseInsensitive=true}.
    private static final ObjectMapper CdrJson = new ObjectMapper()
            .registerModule(new JavaTimeModule())
            .configure(MapperFeature.ACCEPT_CASE_INSENSITIVE_PROPERTIES, true)
            .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);

    private final CdrProcessor _cdrProcessor;

    @Inject
    public ProcessCdrBatchHandler(CdrProcessor cdrProcessor) {
        this._cdrProcessor = cdrProcessor;
    }

    public CdrBatchResult Handle(CdrBatchRequest request) {
        List<cdr> cdrs;
        try {
            cdrs = new ArrayList<>(request.getCdrsJsonCount());
            for (String json : request.getCdrsJsonList())
                cdrs.add(CdrJson.readValue(json, cdr.class));
        } catch (Exception ex) {
            return CdrBatchResult.newBuilder().setError("cdr json parse error: " + ex.getMessage()).build();
        }

        CdrProcessingResult result = _cdrProcessor.ProcessBatch(request.getTenant(), cdrs);
        if (result.Batch() == null)
            return CdrBatchResult.newBuilder()
                    .setCommitted(result.Committed())
                    .setError(result.Error() != null ? result.Error() : "")
                    .build();

        var b = result.Batch();
        return CdrBatchResult.newBuilder()
                .setCommitted(true)
                .setRated(b.Rated().size())
                .setErrored(b.Errored().size())
                .setCdrsWritten(b.CdrsWritten())
                .setChargeablesWritten(b.ChargeablesWritten())
                .setCdrErrorsWritten(b.CdrErrorsWritten())
                .setTotalCharged(b.TotalCharged().doubleValue())
                .build();
    }
}
